### ISSUE TITLE
Any guidelines to a sequence labeling task using keras?

### ISSUE THREAD

**Issue by o0laika0o** [fchollet/keras#2934]
Hi everyone,

I have worked on theano for a period but a newbie to keras. Currently I am going to implement a sequence labeling task using keras. I have read some codes in the example dir but still feel confused about the architecture of keras models on sequence labeling.

Task:
Tagging each sentence in a dialogue with 1 of the pre-defined categories (say 0,1,2,3), and we might call the task as dialogue act tagging (classification).
i.e.
Batch 1:
Q: sentence1 -> 1
A: sentence2  ->2
Q: sentence3  ->0
A: sentence4  ->3
...
Each dialogue (of variable length) is compromised by several sentences, and the sentences thus form a sequence (a sequence of sentences). (Am I clear?)

My preliminary idea is to use a pipeline analogous to [https://github.com/fchollet/keras/blob/master/examples/imdb_cnn_lstm.py], with a CNN and an RNN (i.e. LSTM) to encode each sentence (i.e. sentence2) and the sequence (i.e. sentence1~4) respectively.

The current plan is just like this:
`model = Sequential()`
 `model.add(Embedding(weights=[embedding_weights])) #the pretrained embeddings` 
 `model.add(Convolution1D) #for each sentence`
 `model.add(MaxPooling1D) #for each sentence`
 `model.add(TimeDistributed(LSTM)) #to wrap all the sentences?`
 `model.add(Activation)`

I notice that TimeDistributed is the function to wrap multiple steps but it seems that I couldn't learn much about it from the given example usage.

Question1: Anyone could provide some guidelines? 
Question2: How should my input data look like (just like below)?
X_train = [[1,2,3], [4,5,6], [7,8,9], [8,7,6]]
y_train = [1,2,0,3]
Question3: There is no need to padding the sequence (not the sentence), right? I remember there is a None parameter.

Thanks in advance.

**Comment by o0laika0o** [comment]
With the help of many previous issues, I finally compile and run the code successfully using some TimeDistributed layers (on Embedding, Convolution, MaxPooling and Dense after LSTM). 

The model is just like:

model = Sequential()
model.add(TimeDistributed(Embedding(weights=[embedding_weights]), input_shape=(maxlen_dialog, maxlen_sent))) #the pretrained embeddings 
model.add(TimeDistributed(Convolution2D)) #for each sentence
model.add(TimeDistributed(MaxPooling2D)) #for each sentence
model.add(LSTM) #without timedistributed
model.add(TimeDistributed(Dense)) #predict each sentence
model.add(Activation)

But I find it runs quite slow, most probably I deem is incorrect usage of Masking. 

I added the Masking layer before LSTM to mask the sequences with all zeros, but it seems that the CNN computation for those masked sequences cannot be avoided since the convolution layer doesn't accept shapes that not fixed and as well it doesn't accept a sequence of variable length.

AFSIK, keras only solves this kind of problem by using padding to sequences or sequences of sequences, and masking them before embedding or rnn layers.(i.e. #40)

And I would be appreciated if someone could provide some sample code or some useful suggestions. @fchollet @sergeyf

**Comment by muzixiang** [comment]
Hi @o0laika0o , I want to do sequence classification just like the first picture in Fig 3 in this article http://arxiv.org/pdf/1411.4389v4.pdf , in this article, they said the output label is the average of the softmax of each time step, do you kown how to do this in keras? is it same to your problem?

**Comment by muzixiang** [comment]
@o0laika0o I also want to know why you use timedistributed Dense?

**Comment by o0laika0o** [comment]
@muzixiang In my problem, each sequence would get a tag (e.g. tag each sentence in a paragraph with a label). The input of the TimeDistributed(Dense) is the output of LSTM along the timeline (many steps), and the output also includes many steps.

About the average you mentioned, I think you can write a lambda layer by yourself or you can refer to the Merge layer by setting the merge mode to ave.

Actually I haven't quite complete my code, and it just can compile and run but very slow. So when you get any progress, I would be appreciated if you share you idea with me.
